File: Soduko_solver/Soduko_solver.py
from typing import List

class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        n = 9  # Board size is always 9x9 for a standard Sudoku.

        # Function to check if placing `ch` in `board[row][col]` is valid.
        def isValid(row, col, ch):
            for i in range(9):
                # Check if `ch` is in the same column or row.
                if board[i][col] == ch or board[row][i] == ch:
                    return False

                # Check if `ch` is in the same 3x3 sub-grid.
                if board[3 * (row // 3) + i // 3][3 * (col // 3) + i % 3] == ch:
                    return False

            return True

        # Recursive function to solve the Sudoku.
        def solve(row, col):
            if row == n:  # If we’ve reached past the last row, the board is solved.
                return True
            if col == n:  # Move to the next row if we’ve reached the last column.
                return solve(row + 1, 0)

            if board[row][col] == ".":  # If the cell is empty, try all possible numbers.
                for i in range(1, 10):
                    if isValid(row, col, str(i)):
                        board[row][col] = str(i)  # Place the number.

                        if solve(row, col + 1):  # Recurse to the next cell.
                            return True
                        board[row][col] = "."  # Backtrack if placing `i` doesn't work.
                return False  # Return False if no valid number can be placed.
            else:
                return solve(row, col + 1)  # Move to the next cell if the current cell is filled.

        solve(0, 0)  # Start solving from the top-left corner.

# Input board.
board = [
    ["5", "3", ".", ".", "7", ".", ".", ".", "."],
    ["6", ".", ".", "1", "9", "5", ".", ".", "."],
    [".", "9", "8", ".", ".", ".", ".", "6", "."],
    ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
    ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
    ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
    [".", "6", ".", ".", ".", ".", "2", "8", "."],
    [".", ".", ".", "4", "1", "9", ".", ".", "5"],
    [".", ".", ".", ".", "8", ".", ".", "7", "9"]
]

# Solve the Sudoku puzzle and print the result.
solution = Solution()
solution.solveSudoku(board)

# Print the solved board.
print(board)

File: Soduko_solver/Soduko_solver.java
import java.util.Arrays;

public class Solution {
    public void solveSudoku(char[][] board) {
        // Start solving the board using recursion and backtracking
        solve(board);
    }

    private boolean solve(char[][] board) {
        int n = 9; // Standard Sudoku board size
        for (int row = 0; row < n; row++) {
            for (int col = 0; col < n; col++) {
                if (board[row][col] == '.') { // Find an empty cell
                    for (char num = '1'; num <= '9'; num++) { // Try all numbers from 1 to 9
                        if (isValid(board, row, col, num)) { // Check if placing num is valid
                            board[row][col] = num; // Place the number

                            // Recurse to solve the rest of the board
                            if (solve(board)) {
                                return true; // Solution found
                            }

                            board[row][col] = '.'; // Backtrack
                        }
                    }
                    return false; // No valid number can be placed, backtrack
                }
            }
        }
        return true; // Board completely solved
    }

    private boolean isValid(char[][] board, int row, int col, char num) {
        for (int i = 0; i < 9; i++) {
            // Check row, column, and 3x3 sub-grid
            if (board[row][i] == num || board[i][col] == num ||
                    board[3 * (row / 3) + i / 3][3 * (col / 3) + i % 3] == num) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        // Input Sudoku board
        char[][] board = {
                {'5', '3', '.', '.', '7', '.', '.', '.', '.'},
                {'6', '.', '.', '1', '9', '5', '.', '.', '.'},
                {'.', '9', '8', '.', '.', '.', '.', '6', '.'},
                {'8', '.', '.', '.', '6', '.', '.', '.', '3'},
                {'4', '.', '.', '8', '.', '3', '.', '.', '1'},
                {'7', '.', '.', '.', '2', '.', '.', '.', '6'},
                {'.', '6', '.', '.', '.', '.', '2', '8', '.'},
                {'.', '.', '.', '4', '1', '9', '.', '.', '5'},
                {'.', '.', '.', '.', '8', '.', '.', '7', '9'}
        };

        Solution solution = new Solution();
        solution.solveSudoku(board);
        System.out.println(Arrays.deepToString(board));